base/pmap.jl

```diff
@@ -0,0 +1,102 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+
+"""
+    pgenerate([::WorkerPool], f, c...) -> iterator
+
+Apply `f` to each element of `c` in parallel using available workers and tasks.
+
+For multiple collection arguments, apply f elementwise.
+
+Results are returned in order as they become available.
+
+Note that `f` must be made available to all worker processes; see
+[Code Availability and Loading Packages](:ref:`Code Availability
+and Loading Packages <man-parallel-computing-code-availability>`)
+for details.
+"""
+function pgenerate(p::WorkerPool, f, c)
+    if length(p) == 0
+        return asyncgenerate(f, c)
+    end
+    batches = batchsplit(c, min_batch_count = length(p) * 3)
+    return flatten(asyncgenerate(remote(p, b -> asyncmap(f, b)), batches))
+end
+
+pgenerate(p::WorkerPool, f, c1, c...) = pgenerate(p, a->f(a...), zip(c1, c...))
+
+pgenerate(f, c) = pgenerate(default_worker_pool(), f, c...)
+pgenerate(f, c1, c...) = pgenerate(a->f(a...), zip(c1, c...))
+
+
+
+"""
+    pmap([::WorkerPool], f, c...) -> collection
+
+Transform collection `c` by applying `f` to each element using available
+workers and tasks.
+
+For multiple collection arguments, apply f elementwise.
+
+Note that `f` must be made available to all worker processes; see
+[Code Availability and Loading Packages](:ref:`Code Availability
+and Loading Packages <man-parallel-computing-code-availability>`)
+for details.
+"""
+pmap(p::WorkerPool, f, c...) = collect(pgenerate(p, f, c...))
+#pmap(f, c...) = pmap(default_worker_pool(), f, c...)
+
+function pmap(f, c...; err_retry=nothing, err_stop=nothing, pids=nothing)
```

Comment:
On one hand, it is arguable that this should be a breaking change because:

 - both the defaults `error_retry=true` and `error_stop=false` are inconsistent with `map`.
 - `err_retry=true` by default is dangerous if `f` has side-effects.
 - most code that uses pmap won't have a huge number of call-sites (so its quick to fix)
 - the "breakage" is limited to error handling, so the risk is low that broken code will persist.

On the other hand, it might make sense to start with a non-breaking change and make the breaking change in a seperate PR, or defer the breakage until later something like this...

```julia
if VERSION < v"0.5.1"

function pmap(f, c...; err_retry=true, err_stop=false, pids=default_worker_pool())

    if err_retry == true
        depwarn("`err_retry=true` is deprecated, " *
                "use `pmap(retry(f), c; err_retry = false)`.", :pmap)
        f = retry(f)
    end

    if err_stop == false
        depwarn("`err_stop=true` is deprecated, " *
                "use `pmap(@catch(f), c; err_stop = true)`.", :pmap)
        f = @catch(f)
    end

    if pids != default_worker_pool()
        depwarn("`pids` is deprecated, use `pmap(::WorkerPool, f, c).", :pmap)
        p = WorkerPool(pids)
    end

    return pmap(p, f, c...)
end

elseif VERSION < v"0.5.2"

function pmap(f, c...; err_retry=nothing, err_stop=nothing)

    if err_retry != nothing || err_stop != nothing
        depwarn("`err_retry` and `err_stop` are deprecated, " *
                "use `pmap(retry(f), c)` or `pmap(@catch(f, c))`.", :pmap)
    end

    return pmap(default_worker_pool(), f, c...)
end

else # VERSION >= "v0.5.2"

pmap(f, c...) = pmap(default_worker_pool(), f, c...)

end
```